### ISSUE TITLE
Migrate all CalculateDerivative consumers to the autodiff tape, then delete it from ILossFunction<T>

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1992]
## Why

`ILossFunction<T>` mandates `CalculateDerivative(Vector<T>, Vector<T>)`, but 7 of the 49 loss
classes cannot honour it and throw `NotSupportedException` because they are differentiated
through the autodiff tape:

`APNet2GeneratorLoss`, `AdversarialLoss`, `ContrastiveLoss`, `FlowLoss`,
`NoiseContrastiveEstimationLoss`, `PerceptualLoss`, `TripletLoss`

An optimizer that dispatches through the interface hits a runtime throw. This was raised as
BLOCKING on #1968 against `APNet2GeneratorLoss`, but it is true of six siblings too, so the
root cause is the interface rather than any one class.

PyTorch has no equivalent method at all: an `nn.Module` loss defines only `forward`, and
gradients come from autograd. A hand-written backward exists only in `torch.autograd.Function`,
where you opt in by defining `forward` and `backward` together. There is no interface obliging
every loss to expose an analytic gradient, so this situation cannot arise there. That is the
end state we want.

## Scope (measured, not estimated)

| | Count |
|---|---|
| Call sites outside `src/LossFunctions/` | **204** |
| Distinct files | **95** |
| Loss classes with a real analytic implementation | 42 of 49 |
| Loss classes that throw | 7 of 49 |

By area: ReinforcementLearning 32 files, tests 24, NeuralNetworks 10, Finance 10,
Regression 4, Audio 4, NER 3, then TimeSeries / Optimizers / Models / MetaLearning /
Interfaces 1 each.

`LossFunctionBase.CalculateLossAndGradientGpu` also depends on it (line 69) as the default
GPU gradient path, so that fallback has to be re-expressed before the member can go.

## Work

1. Provide a tape-based gradient helper that returns a `Vector<T>` for a given
   `(loss, predicted, actual)`, so consumers have a drop-in replacement.
2. Migrate the 204 call sites, area by area, largest first.
3. Re-express `LossFunctionBase.CalculateLossAndGradientGpu` on the tape.
4. Delete `CalculateDerivative` from `ILossFunction<T>`, `LossFunctionBase<T>`, and the 42
   analytic implementations.
5. Delete the 7 `NotSupportedException` throws, which then have nothing to throw from.

Its own branch and its own review -- too large and too cross-cutting to land inside a
review-remediation slice.

## Evidence

Audit run on `split/1789-03-models-lossfunctions`:

    git grep -c "\.CalculateDerivative(" -- src tests | grep -v "^src/LossFunctions/"
    # 204 call sites across 95 files
